### ISSUE TITLE
Floating point exception when computing gradient from empty boolean-mask slicing

### ISSUE THREAD

**Issue by BarclayII** [pytorch/pytorch#16635]
## 🐛 Bug

A floating point exception is thrown when one mask-slices a two-dimensional tensor with a one-dimensional all-false boolean tensor.

## To Reproduce

```
In [1]: import torch                                                                                                                                                                                                                                                              

In [2]: x = torch.randn(10, 8).requires_grad_()                                                                                                                                                                                                                                   

In [3]: y = x[(x.sum(1) > 0) & (x.sum(1) < 0)]    # guaranteed to be all false                                                                                                                                                                                                    

In [4]: z = y.sum()                                                                                                                                                                                                                                                               

In [5]: z.backward()                                                                                                                                                                                                                                                              
Floating point exception (core dumped)
```

## Expected behavior

PyTorch shouldn't crash, and preferably the gradient of x should be zero.

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Fedora release 29 (Twenty Nine)
GCC version: (GCC) 8.2.1 20181215 (Red Hat 8.2.1-6)
CMake version: version 3.12.1

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 10.0.130
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7.0.5
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2018.0.3                      1  
[conda] mkl-service               1.1.2            py36h90e4bf4_5  
[conda] mkl_fft                   1.0.6            py36h7dd41cf_0  
[conda] mkl_random                1.0.1            py36h4414c95_1  
[conda] pytorch                   1.0.0           py3.6_cuda9.0.176_cudnn7.4.1_1    pytorch

## Additional context

`masked_select` with an all-false tensor works fine.

**Comment by zou3519** [comment]
I can't reproduce this on master or 1.0.0 on my machine

**Comment by BarclayII** [comment]
Weird.  It occurs on Python 3.5 and pip installation for me as well.

**Comment by BarclayII** [comment]
I tried building the master branch from source and it seems that the problem no longer exists.
